File: java/src/player/property/item/Item.java
package player.property.item;

import damage.Damage;
import player.property.PlayerProperty;
import player.property.item.items.*;
import graphics.ImageID;
import misc.Debug;

public abstract class Item extends PlayerProperty
{
	private byte id;
	private static Item[] items;

	static
	{
		items = new Item[]
		{
			new HealthRing(),
			new DamageRing()
		};

		for (byte i = 0; i < items.length; i++)
		{
			items[i].id = i;
		}
	}

	// for sub
	public abstract int getMassStat();
	public float getHealthStat() { return 0; }
	public float getRegenerationStat() { return 0; }
	public Damage getDamageStat() { return new Damage(0, 0, 0); }

	public static Item[] getAllItems() { return items; }

	// getter
	public static Item getByID(byte id)
	{
		if (id >= 0 && id < items.length)
		{
			return items[id];
		}
		return null;
	}

	@Override public final byte getID() { return id; }
}

File: cpp/src/player/property/item/Item.cpp
#include "Item.hpp"

#include "../../../misc/Converter.hpp"
#include "../../../misc/Debug.hpp"
#include "items/HealthRing.hpp"

std::vector<Item*> Item::items;

void Item::init()
{
	items.push_back(new HealthRing());

	for (int i = 0; i < items.size(); i++)
	{
		items[i]->setID(i);
	}
}

void Item::uninit()
{
	items.clear();
}

Item* Item::get(int id)
{
	Debug::warnIf(id < 0 || id >= getAmount(), "Item::get(): id(" + Converter::intToString(id) + ") out of range");
	return items[id];
}

int Item::getAmount()
{
	return items.size();
}

const std::vector<Item*> Item::getAllItems()
{
	return items;
}

void Item::setID(char id)
{
	this->id = id;
}

char Item::getID() const
{
	return id;
}

int Item::getMassStat() const
{
	return 0;
}

float Item::getHealthStat() const
{
	return 0;
}
